pkg/zstdchunked/storage.go

```diff
@@ -0,0 +1,977 @@
+	maps := make(map[string]map[string]*FileMetadata)
+
+	for layerID, v := range layersMetadata {
+		r := make(map[string]*FileMetadata)
+		for i := range v {
+			r[v[i].Digest] = &v[i]
+		}
+		maps[layerID] = r
+	}
+	return maps
+}
+
+func getLayersCache(store storage.Store) (map[string][]FileMetadata, map[string]string, error) {
+	allLayers, err := store.Layers()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	layersMetadata := make(map[string][]FileMetadata)
+	layersTarget := make(map[string]string)
+	for _, r := range allLayers {
+		manifestReader, err := store.LayerBigData(r.ID, bigDataKey)
+		if err != nil {
+			continue
+		}
+		defer manifestReader.Close()
+		manifest, err := ioutil.ReadAll(manifestReader)
+		if err != nil {
+			return nil, nil, err
+		}
+		var toc TOC
+		if err := json.Unmarshal(manifest, &toc); err != nil {
+			continue
+		}
+		layersMetadata[r.ID] = toc.Entries
+		target, err := store.DifferTarget(r.ID)
+		if err != nil {
+			return nil, nil, err
+		}
+		layersTarget[r.ID] = target
+	}
+
+	return layersMetadata, layersTarget, nil
+}
+
+func MakeZstdChunkedDiffer(ctx context.Context, store storage.Store, blobSize int64, annotations map[string]string, iss ImageSourceSeekable) (*ChunkedZstdDiffer, error) {
+	manifest, err := ReadZstdChunkedManifest(iss, blobSize, annotations)
+	if err != nil {
+		return nil, err
+	}
+	layersMetadata, layersTarget, err := getLayersCache(store)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ChunkedZstdDiffer{
+		stream:         iss,
+		manifest:       manifest,
+		ctx:            ctx,
+		layersMetadata: layersMetadata,
+		layersTarget:   layersTarget,
+	}, nil
+}
+
+func findFileInOtherLayers(file FileMetadata, root string, dirfd int, layersMetadata map[string]map[string]*FileMetadata, layersTarget map[string]string, missingDirsMode, mode os.FileMode) (*os.File, int64, error) {
+	// this is ugly, needs to be indexed
+	for layerID, checksums := range layersMetadata {
+		m, found := checksums[file.Digest]
+		if !found {
+			continue
+		}
+
+		source, ok := layersTarget[layerID]
+		if !ok {
+			continue
+		}
+
+		sourceFile, err := securejoin.SecureJoin(source, m.Name)
+		if err != nil {
+			continue
+		}
+
+		dstFile, written, err := copyFileContent(sourceFile, file.Name, root, dirfd, missingDirsMode, mode)
+		if err != nil {
+			continue
+		}
+		return dstFile, written, nil
+	}
+	return nil, 0, nil
+}
+
+// findFileOnTheHost checks whether the requested file already exist on the host and copies the file content from there if possible.
+// It is currently implemented to look only at the file with the same path.  Ideally it can detect the same content also at different
+// paths.
+func findFileOnTheHost(file FileMetadata, root string, dirfd int, missingDirsMode, mode os.FileMode) (*os.File, int64, error) {
+	hostContentDigester := digest.Canonical.Digester()
+	hostContentChecksum := hostContentDigester.Hash()
+
+	sourceFile := filepath.Clean(filepath.Join("/", file.Name))
```

Comment:
Note that this does not detect symlinks at all, e.g. `/usr/tmp`.

The whole idea of using host files still worries me quite a bit.

Reply:
Yeah, this seems like a good thing to do as a separate PR.